packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -0,0 +1,314 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:vm_service/vm_service.dart';
+
+import '../../globals.dart';
+
+/// Responsible for managing the debug state of the app.
+class DebuggerController {
+  VmService _service;
+
+  StreamSubscription<Event> _debugSubscription;
+
+  IsolateRef isolateRef;
+  List<ScriptRef> scripts;
+
+  final Map<String, Script> _scriptCache = <String, Script>{};
+
+  final _isPaused = ValueNotifier<bool>(false);
+  ValueListenable<bool> get isPaused => _isPaused;
+
+  final _hasFrames = ValueNotifier<bool>(false);
+  ValueNotifier<bool> _supportsStepping;
+  ValueListenable<bool> get supportsStepping {
+    return _supportsStepping ??= () {
+      final notifier = ValueNotifier<bool>(_isPaused.value && _hasFrames.value);
+      void update() {
+        notifier.value = _isPaused.value && _hasFrames.value;
+      }
+
+      _isPaused.addListener(update);
+      _hasFrames.addListener(update);
+      return notifier;
+    }();
+  }
+
+  Event lastEvent;
+
+  final _breakpoints = ValueNotifier<List<Breakpoint>>([]);
+  ValueListenable<List<Breakpoint>> get breakpoints => _breakpoints;
```

Comment:
this was mentioned on a recent PR, but we should consider creating a custom ListNotifier subclass that calls `notifyListeners` when the contents of the list change.  `notifyListeners` is a protected method so it would be nice to not have to violate the rules of "protected" to ensure listeners are being fired at the right time.